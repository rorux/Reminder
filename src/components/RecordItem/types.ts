import React from 'react';
import { TRecordAccordion } from '@components/RecordsAccordion/types';

export type TRecordItemProps = {
  record: TRecordAccordion;
  handleChange: (panel: string) => (event: React.SyntheticEvent, isExpanded: boolean) => void;
  expanded: string | false;
  handleDelete: (event: React.MouseEvent<HTMLElement>, id: string) => void;
};
