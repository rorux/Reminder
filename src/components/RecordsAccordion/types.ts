export type TRecordAccordion = {
  period: string;
  title: string;
  id: string;
  comment: string;
};

export type TRecordsAccordionProps = {
  recordListRender: Array<TRecordAccordion>;
  handleDelete: (event: React.MouseEvent<HTMLElement> | null, id: string) => void;
};
