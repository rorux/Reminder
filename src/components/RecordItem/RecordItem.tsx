import React from 'react';
import { Redirect } from 'react-router-dom';
import Accordion from '@mui/material/Accordion';
import AccordionDetails from '@mui/material/AccordionDetails';
import AccordionSummary from '@mui/material/AccordionSummary';
import Typography from '@mui/material/Typography';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import Stack from '@mui/material/Stack';
import Button from '@mui/material/Button';
import Confirm from '@components/Confirm';
import { TRecordItemProps } from './types';

const RecordItem: React.FC<TRecordItemProps> = ({
  record,
  expanded,
  handleChange,
  handleDelete,
}) => {
  const [isOpenConfirm, setIsOpenConfirm] = React.useState(false);
  const [isEdit, setIsEdit] = React.useState(false);

  if (isEdit) return <Redirect to={`/edit/${record.id}`} />;

  return (
    <Accordion
      expanded={expanded === `panel${record.id}`}
      onChange={handleChange(`panel${record.id}`)}
      key={record.id}
    >
      <AccordionSummary
        expandIcon={<ExpandMoreIcon />}
        aria-controls={`panel${record.id}bh-content`}
        id={`panel${record.id}bh-header`}
      >
        <Typography sx={{ width: '33%', flexShrink: 0 }}>{record.period}</Typography>
        <Typography sx={{ color: 'text.secondary' }}>{record.title}</Typography>
      </AccordionSummary>
      <AccordionDetails>
        <Typography>{record.comment}</Typography>
        <Stack direction="row" spacing={2} sx={{ mt: 2 }}>
          <Button
            data-test="edit-button"
            variant="contained"
            color="warning"
            sx={{ boxShadow: 0 }}
            onClick={() => setIsEdit(true)}
          >
            Изменить
          </Button>
          <Button
            data-test="delete-button"
            variant="contained"
            color="error"
            sx={{ boxShadow: 0 }}
            onClick={() => setIsOpenConfirm(true)}
          >
            Удалить
          </Button>
          <Confirm
            open={isOpenConfirm}
            setOpen={setIsOpenConfirm}
            handler={handleDelete}
            id={record.id}
          />
        </Stack>
      </AccordionDetails>
    </Accordion>
  );
};

export default RecordItem;
