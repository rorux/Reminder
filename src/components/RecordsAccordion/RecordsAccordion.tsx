import React from 'react';
import Accordion from '@mui/material/Accordion';
import AccordionDetails from '@mui/material/AccordionDetails';
import AccordionSummary from '@mui/material/AccordionSummary';
import Typography from '@mui/material/Typography';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import { TRecordsAccordionProps } from './types';

const RecordsAccordion: React.FC<TRecordsAccordionProps> = ({ recordListRender }) => {
  const [expanded, setExpanded] = React.useState<string | false>(false);

  const handleChange = (panel: string) => (event: React.SyntheticEvent, isExpanded: boolean) => {
    setExpanded(isExpanded ? panel : false);
  };

  return (
    <>
      {recordListRender.map((record) => {
        return (
          <Accordion
            expanded={expanded === `panel${record.id}`}
            onChange={handleChange(`panel${record.id}`)}
            key={record.id}
          >
            <AccordionSummary
              expandIcon={<ExpandMoreIcon />}
              aria-controls={`panel${record.id}bh-content`}
              id={`panel${record.id}bh-header`}
            >
              <Typography sx={{ width: '33%', flexShrink: 0 }}>{record.period}</Typography>
              <Typography sx={{ color: 'text.secondary' }}>{record.title}</Typography>
            </AccordionSummary>
            <AccordionDetails>
              <Typography>{record.comment}</Typography>
            </AccordionDetails>
          </Accordion>
        );
      })}
    </>
  );
};

export default RecordsAccordion;
