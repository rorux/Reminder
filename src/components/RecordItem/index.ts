import RecordItem from './RecordItem';

export default RecordItem;
