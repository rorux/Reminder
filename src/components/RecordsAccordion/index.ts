import RecordsAccordion from './RecordsAccordion';

export default RecordsAccordion;
