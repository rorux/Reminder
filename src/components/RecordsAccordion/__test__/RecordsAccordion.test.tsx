import React from 'react';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Enzyme, { shallow } from 'enzyme';
import RecordsAccordion from '../RecordsAccordion';
import RecordItem from '@components/RecordItem';

Enzyme.configure({ adapter: new Adapter() });

const recordListRender = [
  {
    id: '1',
    period: 'Ежемесячно',
    title: 'title of monthly remind',
    comment:
      '8-го числа. В случае выпадения на выходной или праздничный день перенос на ближайший рабочий день ПОСЛЕ наступления срока.',
  },
];

describe('RecordsAccordion component', () => {
  let component: Enzyme.ShallowWrapper;
  let setState: any;
  let useStateSpy;
  beforeEach(() => {
    setState = jest.fn();
    useStateSpy = jest.spyOn(React, 'useState');
    // @ts-ignore
    useStateSpy.mockImplementation((init) => [init, setState]);
    component = shallow(
      <RecordsAccordion recordListRender={recordListRender} handleDelete={jest.fn()} />
    );
  });

  it('Should render RecordsAccordion component', () => {
    expect(component).toMatchSnapshot();
  });

  it('Should be called "useState"', () => {
    const func = component.find(RecordItem).props().handleChange('123');
    func(null, false);
    expect(setState).toHaveBeenCalledWith(false);

    func(null, true);
    expect(setState).toHaveBeenCalledWith('123');
  });
});
