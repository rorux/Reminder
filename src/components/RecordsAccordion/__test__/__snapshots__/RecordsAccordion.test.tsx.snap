// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordsAccordion component Should render RecordsAccordion component 1`] = `
<Fragment>
  <RecordItem
    expanded={false}
    handleChange={[Function]}
    handleDelete={[MockFunction]}
    key="1"
    record={
      Object {
        "comment": "8-го числа. В случае выпадения на выходной или праздничный день перенос на ближайший рабочий день ПОСЛЕ наступления срока.",
        "id": "1",
        "period": "Ежемесячно",
        "title": "title of monthly remind",
      }
    }
  />
</Fragment>
`;
